### PR TITLE
Combining CLIPA-v2 and SigLIP (both big_vision based) models

### PR DESCRIPTION
**Pull request by rwightman**

### CODE REVIEW

**Comment by rwightman** [comment]
Almost there on this one, it got a bit epic. Still some more cleanup to do but want to get this in for more eyes.

Main goal is to support SigLIP pretrained weights and CLIPA-v2 (I don't think supporting v1 is worth it at this point)

There is associated cleanup and additions to support
* HF tokenization with extra post-processing
* Different image interpolation and resizing modes, add support for setting via pretrained configs
* Builtin TextTransformer pooling options (first/last/argmax)
* Builtin ViT pooling options (tok/avg)
  * avg pool is always on spatial tokens, even though some models were trained to incl cls, still works well so no going to support both anymore
* Tried to merge ViT pool cleanly with the attentional pooling for CoCa and provide a path to fix the bugs in that (cascade or parallel)
* Allow final LN in ViT to be applied before or after the pool (differs for some models, CLIPA is after avg pool, SigLIP  before avg pool)
* Remove input patchnorm for future maintenance ease

For @zw615 
* I did not find any noteworthy difference btw tanh approx and normal GELU. This matches my past experience porting weights from jax trained models. I'll leave that out for simplicity
* The TF tokenizer is insanely slow as used and not acceptable as a dep, HF bert uncased appears to match with an added post-processing step

I still have more testing to do, I've run through all SigLIP weights, and ~50% of the v2 CLIPA. At the large end, both the SO400M SigLIP and the new G/14 CLIPA check out at a hair past 83% top-1 

There is a drop of ~0-0.2 in the zero-shot scores from jax versions, this fits past experience.

Need to do some more testing, push weights to hub, CoCa is probably broken right now.

**Comment by rwightman** [comment]
@zw615 do you want to set up and organization and HF hub so we can push under something official? Or should I do it into the timm organization myself? It's much easier to get the weights on the hub.

If you do set up an org I can provide some command lines to push once I have everything else ready.

**Comment by rom1504** [comment]
can you fix the merge conflict so tests will run please?

File: src/open_clip/big_vision.py
Comment (rwightman) [reply src/open_clip/big_vision.py:9]: @rom1504 it works for big vision siglip weights -> timm + builtin text models only, I don't see why it wouldn't be stable, it has nothing to do with the code revision, it's only the weight files that are relevant. This wouldn't work for their lit models, but could be extended, etc.

I could drop this code after I convert and push to HF hub, but it could be useful reference or for future models. It's isolated to checkpoint loading with a npz/npy filename and has no extra deps.

**Comment by rwightman** [comment]
> can you fix the merge conflict so tests will run please?

Pretty sure it's broken at this point, have some more things to do before I'd expect a pass. Will merge main tomorrow. Would like bigger picture feedback/comments early, regardless of tests.

**Comment by rwightman** [comment]
Also @gabrielilharco if you have any comments...

**Comment by rwightman** [comment]
@gpucce while you're poking around here, there are changes impacting CoCa in this PR (and probably some of your future adds) ... just a heads up, I put a sketch of an attentional pool as desribed in the paper in there, while keeping bwd compat

**Comment by gabrielilharco** [comment]
really cool @rwightman! At a high level the design looks good to me. Since a lot changed and some of it affects existing models, we should make sure that the performance of all pretrained models we support remains the same after the changes

**Comment by gpucce** [comment]
> @gpucce while you're poking around here, there are changes impacting CoCa in this PR (and probably some of your future adds) ... just a heads up, I put a sketch of an attentional pool as desribed in the paper in there, while keeping bwd compat

thanks, will try and keep it all together, there was another small PR trying to address this, #551, fixing another small issue, don´t know if you ever had time to have a look

**Comment by rom1504** [comment]
It's a lot of changes but looks ok to me. Would definitely be safer to merge (and potentially revert) in small PRs bringing more scoped changes 

My main worry is I can't tell from the code if things still work.

The tests partially cover inference still working the same. Maybe it needs to be expanded to new models

What about training, how do we check?

File: src/open_clip/big_vision.py
Comment (rom1504) [reply src/open_clip/big_vision.py:9]: Ok you're saying this will work for some specific big vision trained checkpoints.

Comment (rom1504) [reply src/open_clip/big_vision.py:9]: Then maybe we could indicate those above the function?

**Comment by rwightman** [comment]
@rom1504 yeah, smaller would be better but clipa and siglip are both inter-related in a non-trivial way since they both come big_vision but with differences that conflict with past decisions made here, and they also need extra (but different) tokenizer support, preprocessing, etc. 

As a counter point, things would have ended up more of a mess if CLIPA had been merged as is, and then someone else decided to add SigLIP on top, and the other fixes got added, etc ... at least this will be pulled together with a goal to leave things a bit better in the end with hopefully a cleaner base to add more

**Comment by rwightman** [comment]
@rom1504 I feel testing inference will be enough here, the training risk should be low, the breaks should all be detectable running inference on pretrained models across all types

File: src/open_clip/big_vision.py
Comment (rwightman) [reply src/open_clip/big_vision.py:9]: sure, will make a comment that it only supports the big_vision SigLIP weights right now, any other big_vision weights (present or future) could be added by expanding the support by checking for various keys in the numpy archive....

**Comment by rom1504** [comment]
Looks like tests are passing

Once this is merged, we can adapt https://github.com/mlfoundations/open_clip/blob/e7b39e4321b2c59e7dc2cdf83cddd7742cd9e6a0/tests/test_inference.py#L44 to make sure siglip models also get into non regression tests

**Review by rom1504 — APPROVED** [review]

**Comment by rom1504** [comment]
I checked the code again and the tests results and overall it lgtm

> Need to do some more testing, push weights to hub, CoCa is probably broken right now.

Does this still need to be done? Looks like tests are indeed not checking coca.

What else needs to be done before merging?

Should we add more to the readme to say what is now newly supported?

**Comment by rwightman** [comment]
@rom1504 
* I just pushed SigLIP weights today after working through a better way to ensure the preprocess cfg passes through from Hub, code based configs, and command line w/o breaking backwards compat
* I'm running a script to test CLIPA weight upload to my personal HF page (https://huggingface.co/rwightman) so that I can run a final verification, and then will pass off to @zw615 so that can be uploaded to a UCSC / VLAA org.
* CoCa still needs a manual test to ensure contrastive and generative behaviour is okay
* I need to run through some existing models before a final test

This has improved handling for nltk import that @JeniaJitsev was complaining about, since he's training maybe he can run an epoch on this branch? :)

**Comment by rwightman** [comment]
> @rwightman I'm happy to eval the new and old models if that's useful

@gabrielilharco that'd be very helpful if you've got it setup. All the SigLIP and CLIPA (v2 only) models should be testable ... I put my temporary rwightman/ CLIP models in there prior to the final move, so can run a eval on them...

**Comment by rwightman** [comment]
Oh right sorry bout that. Needs main branch of timm that was just merged
today!

**Comment by gabrielilharco** [comment]
Thanks @rwightman. I'm still getting last error (`RuntimeError: The size of tensor a (77) must match...`) when using the latest commit from timm (`e1e7cf5275600ea74d7b6ed80cc105701bddefdb`)

The good news is that I was able to run evals on all other pretrained models and all numbers match 🥳

**Comment by gabrielilharco** [comment]
yes, everything from `open_clip.list_pretrained()`

**Comment by gabrielilharco** [comment]
@rwightman one option is to propagate the context_length from the configs to the tokenizer initialization [here](https://github.com/mlfoundations/open_clip/blob/72b75bde2ea8427dcb87119ce570c0f50f09c99a/src/open_clip/tokenizer.py#L264), and then in __call__ we use that value if no `context_length` arg is used, otherwise keep the one from the init

**Comment by rwightman** [comment]
@gabrielilharco yeah, thought about that option... so options are
* make the caller grab context length from model and pass in on tokenize() call (or bind with partial themselves as the open_clip scripts do in this PR)
* add an arg to __init__ to pass the context len on construction, store as an attribute (can usually be extracted from model config on creation) and use that in the __call__(), make the arg in the __call__() method None and override the class attribute if set
* have the factory function bind the context_len to tokenizer `partial(tokenizer, context_len=x)` before returning, but this requires an extra bool to allow the tokenizer to be returned unbound, as sometimes we want the tokenizer directly to modify it or save_pretrained() etc

**Comment by gabrielilharco** [comment]
The last two seem cleaner to me since there is no need to change downstream code

**Comment by rwightman** [comment]
@gabrielilharco okay! reworked to use a class attribute for context_length with an optional override via the fn argument.

Also fixed up the CLIPA masking, fixed bug reported in #680 and provided a shuffling vs ordered version of the random reduction masking.

**Comment by rwightman** [comment]
So I think this should be almost ready. @gpucce do you have a quick text gen test that can be run for CoCa to compare against main? Looks like the zero-shot classification was verified in the run @gabrielilharco did.

**Comment by gpucce** [comment]
> So I think this should be almost ready. @gpucce do you have a quick text gen test that can be run for CoCa to compare against main? Looks like the zero-shot classification was verified in the run @gabrielilharco did.

@rwightman, I can run fine-tuning and see if perf stays the same, is it too late if I start it now and tell you the results tomorrow morning? it is a bit later for me to wait now.

**Comment by rwightman** [comment]
> > So I think this should be almost ready. @gpucce do you have a quick text gen test that can be run for CoCa to compare against main? Looks like the zero-shot classification was verified in the run @gabrielilharco did.
> 
> @rwightman, I can run fine-tuning and see if perf stays the same, is it too late if I start it now and tell you the results tomorrow morning? it is a bit later for me to wait now.

That'd be fine if started now.

**Comment by gpucce** [comment]
@rwightman it looks like captioning is broken, tomorrow in the afternoon I can try and see if I find the issue, right now it is too late

**Comment by rwightman** [comment]
@gpucce kk, I most likely broke something in the pooling part...

**Comment by gabrielilharco** [comment]
> @gabrielilharco okay! reworked to use a class attribute for context_length with an optional override via the fn argument (so backwards compat behaviour is maintained).

Nice @rwightman! Looking great. I'm able to run evals now for all new models, will send results once they're done

**Comment by rwightman** [comment]
@gpucce think I found it, the LN wasn't being applied properly with the attn_pool set, have to do it differently for the legacy (current) CoCa and future double pool options...

**Comment by gpucce** [comment]
@rwightman it seems there is still smth different for the captions, let me try and have a better look at the outputs more deeply

**Comment by rwightman** [comment]
Oh fun, yeah thought that last change would bring the CoCa ViT outputs to match old ones, they would have been different (for the tokens output) before, the pooled output should have been the same...

**Comment by rwightman** [comment]
FWIW, it's probably worth checking CoCa on the main branch to ensure it was in a good state there too, in case there was a prior regression we didn't pick up

**Comment by gpucce** [comment]
@rwightman until booster is up (tomorrow) I can't run the evals again because the gpu I have where I am is filled right now and also leonardo is under maintenance currently I am checking instances by hand

**Comment by gabrielilharco** [comment]
> FWIW, it's probably worth checking CoCa on the main branch to ensure it was in a good state there too, in case there was a prior regression we didn't pick up

@rwightman good point, I'm running some evals on main now and they are matching the numbers from [d7542e4](https://github.com/mlfoundations/open_clip/pull/660/commits/d7542e4baaa255080b0202d9f08339f0e768655c) so far

**Comment by rwightman** [comment]
@gabrielilharco @gpucce okay, so I did have another issue in the text pool, the 'tokens' included the cls token which would impact generation, but not zero-shot eval. Fixed that in the very latest. But for both of the previous fixes, the broken path was in the 'tokens' output from vision or text encoder, so zero-shot should have been consistent with main.

**Comment by rwightman** [comment]
@gabrielilharco nice eval looking good, but double checked that bigG CLIPA, should be 83, looks like I mixed up a config value, durr... confirming

**Comment by gpucce** [comment]
@rwightman eval for already fine-tuned model is consistent with main before, I am running fine-tuning

**Comment by gpucce** [comment]
> @rwightman eval for already fine-tuned model is consistent with main, I am running fine-tuning

fine-tuning seems fine too

**Comment by rwightman** [comment]
@gpucce nice, thanks!

okay, so are we ready to merge? @gabrielilharco 

* the bigG CLIPA needs updated eval after last fix to config
* I can get CLIPA HF hub models moved to a UCSC related location @zw615 (I think I can have someone move the ones I've already uploaded to a different org, and then the README can be updated as desired). Also, the bigG 224x224 model is missing, has same link as the 336? 
*

**Comment by gabrielilharco** [comment]
@rwightman all good from my side! I'm getting 83.09 for bigG CLIPA on ImageNet. Full evals are still running (it's a big slow since the model is so big),  I'll update the numbers once their done

**Review by gabrielilharco — APPROVED** [review]

**Comment by rwightman** [comment]
k, I will merge now so we can get changes and various fixes in now. 

Maybe hold off on a versioned release for another day or two to give a few early adopters chance to try main branch? 

We should still add some text to README / PRETRAINED.MD about CLIPA-v2 and SigLIP but can think about that one...

**Comment by rom1504** [comment]
Great!

I think as follow up to merging it would be nice to
- include clippa and siglip in the regression tests so they keep working
- same for coca
- add a regression test for the caption generation of coca

The first 2 are just about enabling these models in the test, no code
change

**Comment by gpucce** [comment]
@rom1504 @rwightman I can add the reg tests for generation

**Comment by zw615** [comment]
Hi @rwightman , sure! I will discuss with and work on this HF organization thing with other co-authors, and should be finish it real quick.

> @gpucce nice, thanks!
> 
> okay, so are we ready to merge? @gabrielilharco
> 
>     * the bigG CLIPA needs updated eval after last fix to config
> 
>     * I can get CLIPA HF hub models moved to a UCSC related location @zw615 (I think I can have someone move the ones I've already uploaded to a different org, and then the README can be updated as desired). Also, the bigG 224x224 model is missing, has same link as the 336?

**Comment by zw615** [comment]
Hi @rwightman , sure! I will discuss with and work on this HF organization thing with other co-authors, and should be finish it real quick.

> @gpucce nice, thanks!
> 
> okay, so are we ready to merge? @gabrielilharco
> 
>     * the bigG CLIPA needs updated eval after last fix to config
> 
>     * I can get CLIPA HF hub models moved to a UCSC related location @zw615 (I think I can have someone move the ones I've already uploaded to a different org, and then the README can be updated as desired). Also, the bigG 224x224 model is missing, has same link as the 336?

**Comment by zw615** [comment]
@rwightman Hi, ross. Could you please move those weights to the _UCSC-VLAA_ organization? (https://huggingface.co/UCSC-VLAA)

**Comment by rwightman** [comment]
Okay, will ask someone to do that on Monday!

**Pull request merged**
